common/djangoapps/entitlements/models.py

```diff
@@ -167,7 +171,7 @@ def update_expired_at(self):
         OR if the policy can no longer be regained AND the policy has been redeemed
         """
         if (not self.expired_at and self.policy.get_days_until_expiration(self) < 0
-                or (self.enrollment_course_run and self.is_entitlement_regainable())):
+                or (self.enrollment_course_run and not self.is_entitlement_regainable())):
```

Comment:
You probably just need the `get_days_until_expiration` check if that method has been updated to take into account the regain logic.

Reply:
I don't think so, because in the is_entitlement_regainable it relies on enrollment_course_run existing otherwise false. Then in this method a course w/o an enrollment would always evaluate to true. But, if a course doesn't have an entitlement, if can't be regained.

Reply:
I'm not sure I understand. `get_days_until_expiration`, when updated as we discussed, should take into account whether or not the entitlement has an associated course run and return the correct number of days until the entitlement actually expires. Therefore, I'm not sure why you need to check anything other than that `get_days_until_expiration(self) < 0` to determine if the entitlement should be expired.